File: backend/src/modules/pedido/dto/create-pedido.dto.ts
import {
  IsString,
  IsNumber,
  IsArray,
  Min,
  IsNotEmpty,
  IsEmail,
  ValidateNested,
  Matches,
} from 'class-validator';
import { Type } from 'class-transformer';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

const TELEFONO_URUGUAY_REGEX = /^(0[1-9]\d{6,7}|09\d{7})$/;

class ItemDto {
  @ApiProperty({ description: 'ID del producto', example: '550e8400-e29b-41d4-a716-446655440000' })
  @IsString()
  readonly productoId: string;

  @ApiProperty({ description: 'Cantidad del producto', example: 2 })
  @IsNumber()
  @Min(1)
  readonly cantidad: number;

  constructor(productoId: string, cantidad: number) {
    this.productoId = productoId;
    this.cantidad = cantidad;
  }
}

export class CreatePedidoDto {
  @ApiProperty({ description: 'Email del comprador', example: 'comprador@email.com' })
  @IsEmail()
  @IsNotEmpty()
  readonly emailComprador: string;

  @ApiProperty({
    description: 'Telefono del comprador (formato Uruguay: 099123456 o 24001234)',
    example: '099123456',
  })
  @IsString()
  @IsNotEmpty()
  @Matches(TELEFONO_URUGUAY_REGEX, {
    message: 'El telefono debe tener formato Uruguay (ej: 099123456 o 24001234)',
  })
  readonly telefonoComprador: string;

  @ApiProperty({ description: 'Items del pedido', type: [ItemDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => ItemDto)
  readonly items: ItemDto[];

  @ApiPropertyOptional({ description: 'Talles de los productos (opcional)', example: ['M', 'L'] })
  @IsString({ each: true })
  readonly talles?: string[];

  constructor(emailComprador: string, telefonoComprador: string, items: ItemDto[]) {
    this.emailComprador = emailComprador;
    this.telefonoComprador = telefonoComprador;
    this.items = items;
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/pedido/dto/CreatePedidoDto.java
package com.sistemapedidos.modules.pedido.dto;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;

import java.util.List;

public record CreatePedidoDto(
        @NotBlank(message = "El email es obligatorio")
        @Email(message = "El email no tiene un formato valido") String emailComprador,
        @NotBlank(message = "El telefono es obligatorio")
        @Pattern(regexp = "^(0[1-9]\\d{6,7}|09\\d{7})$",
                message = "El telefono debe tener formato Uruguay (ej: 099123456 o 24001234)") String telefonoComprador,
        @NotNull(message = "El carrito esta vacio")
        @NotEmpty(message = "El carrito esta vacio")
        @Valid List<ItemDto> items,
        List<String> talles) {

    public record ItemDto(
            @NotBlank(message = "El ID del producto es obligatorio") String productoId,
            @NotNull(message = "La cantidad es obligatoria")
            @Min(value = 1, message = "La cantidad debe ser mayor a cero") Integer cantidad) {
    }
}
